### PR TITLE
Add graph store HTTP protocol

### PR DESCRIPTION
**Pull request by cecton**
Specifications at https://www.w3.org/TR/sparql11-http-rdf-update/#graph-management

### CODE REVIEW

File: aiosparql/client.py [hunk 1]
```diff
@@ -178,6 +185,37 @@ def _pretty_print_query(self, query: str) -> str:
                     explanation=explanation)
             return await resp.json()
 
+    def _crud_request(self, method, graph=None, data=None):
```
Comment (asvetlov) [inline aiosparql/client.py]: What is `data` parameter for?

Comment (asvetlov) [inline aiosparql/client.py]: Looks it it's never used

Comment (cecton) [inline aiosparql/client.py]: HTTP body.

The graph store protocol is basically a crud endpoint where you post/delete/get raw data in any format you want (RDF XML, Turtle, etc...). 

Just realized now that there is no parameter to set the format. Maybe I should start adding integration tests.

Comment (cecton) [inline aiosparql/client.py]: Oh I see. Yes integration test would be very useful at this point.

File: aiosparql/client.py [hunk 2]
```diff
@@ -178,6 +185,57 @@ def _pretty_print_query(self, query: str) -> str:
...
+
+    def head(self, *, format: str, graph: Optional[IRI] = None):
+        return self._crud_request("HEAD", graph=graph, accept=format)
+
+    async def patch(self, data: Union[bytes, IOBase], *,
+                    graph: Optional[IRI] = None):
+        async with self._crud_request("PATCH", graph=graph, data=data,
+                                      accept="application/json") as resp:
+            resp.raise_for_status()
+            return await resp.json()
```
Comment (cecton) [inline aiosparql/client.py]: I'm not sure about this code. According to the specs, the PATCH method is like an endpoint for SPARQL 1.1 update:

> SPARQL 1.1 Update can be used as a patch document. 

I suppose that without being able to really test it I should simply remove the code for now or raise NotImplementedError.

File: aiosparql/client.py [hunk 3]
```diff
@@ -178,6 +185,57 @@ def _pretty_print_query(self, query: str) -> str:
...
+            resp.raise_for_status()
+
+    async def post(self, data: Union[bytes, IOBase], *, format: str,
+                   graph: Optional[IRI] = None):
+        async with self._crud_request("POST", graph=graph, data=data,
+                                      content_type=format) as resp:
+            resp.raise_for_status()
+
+    def head(self, *, format: str, graph: Optional[IRI] = None):
+        return self._crud_request("HEAD", graph=graph, accept=format)
```
Comment (cecton) [inline aiosparql/client.py]: I'm not sure about this code. According to the specs, the HEAD method does the exact same thing than GET but does not return content. So I suppose the format should be supplied by the client...

> the HTTP HEAD method is identical to GET except that the server MUST NOT return a message-body in the response

File: tests/integration/test_client.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    async def test_crud(self):
+        await self.client.put(sample_data, format=sample_format)
+        await self.client.delete()
+        await self.client.post(sample_data, format=sample_format)
+        async with self.client.get(format="text/turtle") as res:
+            self.assertEqual(res.status, 200)
+            text = await res.text()
+            self.assertIn("@prefix", text)
+        async with await self.client.head(format="text/turtle") as res:
+            self.assertIn(res.status, (200, 501))
```
Comment (cecton) [inline tests/integration/test_client.py]: I love doing integration test with a server that does not implement everything...

File: tests/integration/test_client.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        try:
+            res = await self.client.patch("DELETE WHERE { ?s ?p ?o }")
+            self.assertIsInstance(res, dict)
+        except aiohttp.ClientResponseError as exc:
+            if exc.code != 501:
+                raise
```
Comment (cecton) [inline tests/integration/test_client.py]: PATCH is apparently not implemented either by Virtuoso.

**Comment by cecton** [comment]
@asvetlov ready for your review :) I will probably release after this one.

File: aiosparql/client.py (hunk 3 above)
Comment (asvetlov) [inline aiosparql/client.py]: Sorry, I don't follow

File: aiosparql/client.py (hunk 2 above)
Comment (asvetlov) [inline aiosparql/client.py]: +1
If you are not sure about implementation -- just drop the method and wait for feature request with real use case.

**Comment by cecton** [comment]
@asvetlov Yes I can spawn a container for virtuoso... would be simpler to use I suppose... but this doesn't look like it belongs to the test file itself. I guess I could make a "ci" directory with a start and stop script. Does that sound good/better/less good?

**Comment by asvetlov** [comment]
I forgot you use `unittest` style for this project.

It's up to you, you are the library author.
But I suggest considering switch to pytest everywhere (I see you've started already by adding `test_client_context`).

After that you could add code like used by aiopg: https://github.com/aio-libs/aiopg/blob/master/tests/conftest.py#L105-L194

It tests DB client against several Postgres versions by providing a fixture for container start and other fixture for making client connected to started DB server.

**Comment by asvetlov** [comment]
Maybe you'll find the trick useful for aiosparql too.

**Comment by cecton** [comment]
@asvetlov I think I will keep the idea for later for a new PR.

I don't mind much having py.test or unittest tests (even a mix of them) as long as the coverage is good and the code is well tested. It's never too late to change and besides this library is very very small because it's SPARQL: this lib only needs to respect the specifications or SPARQL and HTTP; the server it connects to doesn't really matter. I have worked with PostgreSQL too and I know for sure this is very very different. Overall, this library is some kind of helper to make HTTP requests (at least for `aiosparql.client`).

Now the syntax helpers are more complicated: in SPARQL every value has a scheme for the type and maybe a language if it's a literal, plus depending on the server, the type you specify may not be the type you will get in the database itself... escaping SPARQL is a hell and that's the main purpose of `aiosparql.escape` and generating nodes too `aiosparql.syntax`.

One thing is missing (somehow deliberately) in this library: parsing the SPARQL result and make internal objects. Most of the libraries I know does it but, to be honest, when you use JSON, I just don't find that very useful. So right now I keep an open door for it for later if someone or myself want to implement something at some point. It's just not my priority.

**Pull request merged**